### PR TITLE
New Community creation

### PR DESCRIPTION
**Pull request by memorisanka**
Hi Guys,
I've added a new community creation form and views.
Now user can see all communities, and add a new community (when logged in).

### CODE REVIEW

File: core/models.py [hunk 1]
```diff
@@ -37,6 +36,8 @@ class Meta:
 
 class Community(GenericModel):
     name = models.CharField(max_length=255)
+    slug = models.SlugField(max_length=255, unique=True, blank=True)
+    author = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, related_name='authored_communities')
```
Comment (jacoor) [inline core/models.py:40]: Duplicate field. Generic model has author.

Comment (jacoor) [inline core/models.py]: Good idea with adding related name

File: core/models.py [hunk 2]
```diff
@@ -159,7 +170,8 @@ def is_saved(self: "Post", user: User) -> bool:
     def get_comments(self: "Post") -> models.QuerySet:
         return self.children.all()
 
-    def get_comment_form(self: "Post") -> CommentForm:
+    def get_comment_form(self: "Post"):
```
Comment (jacoor) [inline core/models.py]: Why this change? I mean, removing type?

File: core/templates/core/community-create.html [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+{% endblock %}
+
+{% block content %}
+<style>
+.hidden {
+    display: none;
+}
+
+</style>
+<script>
```
Comment (jacoor) [inline core/templates/core/community-create.html]: I don't see it being used. If it please point me to the place, if not, please remove.

File: core/tests/test_forms.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    form = CommunityForm(data=form_data)
+
+    assert form.is_valid()
+
+    community = form.save(commit=True)
+
+    expected_slug = slugify(form_data["name"])
+    assert community.slug == expected_slug
+
+    community.save()
```
Comment (jacoor) [inline core/tests/test_forms.py]: This save seems unnecessary here, no? Community is created in line 17

**Review by jacoor — CHANGES_REQUESTED** [review]

File: core/models.py (hunk 2 above)
Comment (memorisanka) [reply core/models.py]: I had a problem with circular imports (models from forms, forms from models), so I moved this import only to import in function. When I moved it, the type throws an error.

File: core/templates/core/community-create.html (hunk 3 above)
Comment (memorisanka) [reply core/templates/core/community-create.html]: You mean all file is not used?

Comment (jacoor) [reply core/templates/core/community-create.html]: No, I mean this js script.

File: core/models.py (hunk 2 above)
Comment (jacoor) [reply core/models.py]: I am aware of it. 
What I pointed out was removing type, not the import.

Comment (memorisanka) [reply core/models.py]: Because when import was inside the function, the typing gave an error. I don't have an idea how to solve it.

Comment (jacoor) [reply core/models.py]: oh, interesting. to keep types in place you can add `-> any:` there. Should solve issue and code style is consistent.

File: core/templates/core/community-create.html (hunk 3 above)
Comment (memorisanka) [reply core/templates/core/community-create.html]: OK, removed. Sorry, I'm not good at frontend and js scripts...

Comment (jacoor) [reply core/templates/core/community-create.html]: Easy, thanks.

File: core/models.py (hunk 2 above)
Comment (memorisanka) [reply core/models.py]: No, ruff doesn't allow type Any here! 

obraz

Comment (jacoor) [reply core/models.py]: image
Sorry, small caps. 
This works in other PR.
Also please remove this TYPE_CHECKING check.

Comment (memorisanka) [reply core/models.py]: Corrected. Thanks :)

**Review by jacoor — DISMISSED** [review]

File: core/forms.py [hunk 5]
```diff
@@ -37,3 +40,15 @@ def __init__(self: "Post", *args: tuple, **kwargs: dict) -> None:
...
         self.fields["title"].required = True
         self.fields["content"].required = True
+
+
+class CommunityForm(forms.ModelForm):
+    class Meta:
+        model = Community
+        fields: typing.ClassVar = ["name"]
+
+    def __init__(self: "CommentForm", *args: list, **kwargs: dict) -> None:
```
Comment (yanazPL) [inline core/forms.py]: `self: "CommentForm"` vs `self: "CommunityForm"`

File: core/tests/test_models.py [hunk 6]
```diff
@@ -228,3 +229,15 @@ def test_get_saved_posts(user: User, post: Post) -> None:
     assert saved_posts.first().post == post
     SavedPost.remove_saved_post(user=user, post=post)
     assert not saved_posts.exists()
+
+
+@pytest.mark.django_db()
+def test_create_community() -> None:
+    user = User.objects.create_user(
+        email="testuser@example.com", nickname="Test Nickname", password=os.getenv("TEST_PASSWORD")
```
Comment (yanazPL) [inline core/tests/test_models.py]: you could use `user` fixture conftest.py - password would be generated on the fly

File: core/forms.py (hunk 5 above)
Comment (memorisanka) [reply core/forms.py]: Oo, thanks. Corrected!

File: core/models.py [hunk 7]
```diff
@@ -43,6 +46,16 @@ class Meta:
     def __str__(self: "Community") -> str:
         return str(self.name)
 
+    def save(self: "Community", *args: list, **kwargs: dict) -> None:
+        if not self.slug:
```
Comment (yanazPL) [inline core/models.py:50]: this method is simple and elegant

File: env-template [hunk 8]
```diff
@@ -18,4 +18,7 @@ EMAIL_HOST_USER = "test@gmail.com"
 EMAIL_HOST_PASSWORD = "password"
 EMAIL_HOST = "smtp.gmail.com"
 EMAIL_HOST_PORT = 587
-EMAIL_USE_TLS = True
\ No newline at end of file
+EMAIL_USE_TLS = True
+
+# Tests variables
+TEST_PASSWORD = "password"
```
Comment (yanazPL) [inline env-template]: with `user` fixture it is not needed

File: core/tests/test_models.py (hunk 6 above)
Comment (memorisanka) [reply core/tests/test_models.py]: Thanks, I have changed this in the code. Good to know!

File: env-template (hunk 8 above)
Comment (memorisanka) [reply env-template]: Removed.

**Comment by yanazPL** [comment]
also wondering if we should have more view tests

File: core/views.py [hunk 9]
```diff
@@ -110,3 +111,38 @@ def post(self: "PostSaveView", request: HttpRequest, pk: int, action_type: str)
...
+        response = super().form_valid(form)
+        CommunityMember.objects.create(
+            community=self.object,
+            user=self.request.user,
+            role=CommunityMember.ADMIN,
+        )
+        return response
+
+    def get_success_url(self: "CommunityCreateView") -> str:
+        return reverse_lazy("community-detail", kwargs={"slug": self.object.slug})
```
Comment (yanazPL) [inline core/views.py:139]: another really minor thing, but we can always improve

https://stackoverflow.com/questions/48669514/difference-between-reverse-and-reverse-lazy-in-django

Comment (jacoor) [reply core/views.py:139]: What is your suggestion? I mean, I don't see any harm in using reverse_lazy. Even more, to keep it really simple, I'd suggest using reverse_lazy everywhere unless code error. This way we don't have to think: is this a function, attribute of a class, CBV of FBV. 
Simple is best.

Comment (yanazPL) [reply core/views.py:139]: Just thought there are reasons why we have two versions of these functions and wanted to use them as intended

Comment (yanazPL) [reply core/views.py:139]: But I get it. One variant less confusion

**Review by jacoor — APPROVED** [review]
